### ISSUE TITLE
Date & time functions fail for constant input in PostgreSQL

### ISSUE THREAD

**Issue by apano-on** [ontop/ontop#484]
### Description

Functions for [date & time](https://www.w3.org/TR/sparql11-query/#func-date-time) fail for PostgreSQL when there is a constant input in the SPARQL query.

### Steps to Reproduce

Run the following query:
```
SELECT ?w WHERE
BIND("1799-12-14T09:00:00"^^xsd:dateTime AS ?death )
BIND("1732-02-22T09:30:00"^^xsd:dateTime AS ?birth )
BIND (YEAR(?death) - YEAR(?birth) AS ?w)
```

**Expected behavior:** Return 67

**Actual behavior:** ```Caused by: org.postgresql.util.PSQLException: ERROR: function pg_catalog.date_part(unknown, unknown) is not unique  Hint: Could not choose a best candidate function. You might need to add explicit type casts.```

**Reproduces how often:** Always

### Attached material
Translated query is:
```SELECT (EXTRACT(YEAR FROM '1799-12-14 09:00:00') - EXTRACT(YEAR FROM '1732-02-22 09:30:00'))```
Translated query should be (tentatively):
```SELECT (EXTRACT(YEAR FROM TIMESTAMP '1799-12-14 09:00:00') - EXTRACT(YEAR FROM TIMESTAMP '1732-02-22 09:30:00'))```

### Versions

Ontop v4.1.1 and v4.2.0-SNAPSHOT

### Additional Information

As a tentative fix, what works is simply overriding the serialization for the date & time functions for constants and adding a cast as attempted in 59b0eca229e02f39f4713c34a269a30a7fd8588d. However, I do not know if there is a better way to handle this issue.

**Comment by bcogrel** [comment]
Hi @apano-on , 

From for reporting the issue. This is the responsibility of the `SQLTermSerializer`. Please consider adding a case of the following switch for DATETIME: https://github.com/ontop/ontop/blob/version4/db/rdb/src/main/java/it/unibz/inf/ontop/generation/serializer/impl/DefaultSelectFromWhereSerializer.java#L418

Best,
Benjamin

**Comment by apano-on** [comment]
Hi @bcogrel ,

Thanks for the hint. I applied a trivial change to the switch accounting for DATETIME, and also added a new DBTermType category for DATE which solved the issue. This was also useful as part of a separate change since it also resolved an issue there 5c0a8d2. Let me know if it is not well aligned (or there is any other issue).

I applied the change only for PostgreSQL and Dremio (where I noticed a similar problem). I tried changing the default serializer and noticed it caused an issue for Denodo, since it did not seem possible to cast a string to a TIMESTAMPTZ (when the input contained a timezone and special handling was necessary and I did not want to test thoroughly Denodo time datatypes). Maybe a more generic change can be applied in the future.

**Comment by bcogrel** [comment]
Hi @apano-on 

Thanks. I added the following comment to your commit https://github.com/ontop/ontop/commit/5c0a8d298d63f0a016bb9f5a7251c5dc1d76da46#r61639519 . It does not change the behavior of your code, it just makes things slightly better factorized.

**Comment by apano-on** [comment]
Hi @bcogrel ,

I applied the suggested simplification with 006fddf. Let me know in case of any other change.

**Comment by bcogrel** [comment]
Thanks @apano-on, it looks good to me.

**Comment by apano-on** [comment]
Issue resolved by changes implemented in #483.
